qt/widgets/instrumentview/src/InstrumentActor.cpp

```diff
@@ -615,50 +627,52 @@ void InstrumentActor::sumDetectorsRagged(QList<int> &dets,
 }
 
 /**
-* Recalculate the detector colors based on the integrated values in
-* m_specIntegrs and
-* the masking information in ....
-*/
+ * Recalculate the detector colors based on the integrated values in
+ * m_specIntegrs and
+ * the masking information in ....
+ */
 void InstrumentActor::resetColors() {
   QwtDoubleInterval qwtInterval(m_DataMinScaleValue, m_DataMaxScaleValue);
-  m_colors.resize(m_specIntegrs.size());
-
   auto sharedWorkspace = getWorkspace();
+  const auto &componentInfo = sharedWorkspace->componentInfo();
+  const auto &detectorInfo = sharedWorkspace->detectorInfo();
   const auto &spectrumInfo = sharedWorkspace->spectrumInfo();
+  auto color = m_colorMap.rgb(qwtInterval, 0);
+  m_colors.assign(componentInfo.size(),
+                  GLColor(qRed(color), qGreen(color), qBlue(color), 1));
 
   IMaskWorkspace_sptr mask = getMaskWorkspaceIfExists();
+  const auto &detectorIDs = detectorInfo.detectorIDs();
+  for (size_t wi = 0; wi < m_specIntegrs.size(); ++wi) {
+    const auto &specDef = spectrumInfo.spectrumDefinition(wi);
+    bool masked = false;
 
-  for (int iwi = 0; iwi < int(m_specIntegrs.size()); iwi++) {
-    size_t wi = size_t(iwi);
-    double integratedValue = m_specIntegrs[wi];
-    try {
-      // Find if the detector is masked
-      const auto &dets = sharedWorkspace->getSpectrum(wi).getDetectorIDs();
-      bool masked = false;
+    for (const auto &det : specDef) {
+      auto detIndex = det.first;
 
-      if (mask) {
-        masked = mask->isMasked(dets);
-      } else {
-        masked = spectrumInfo.hasDetectors(wi) && spectrumInfo.isMasked(wi);
-      }
+      if (mask)
+        masked = mask->isMasked(detectorIDs[detIndex]);
```

Comment:
Idea being that if any detectors in the spectrum are masked, then for colouring they should all be treated as masked?